src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/DocumentMapping/RemoteDocumentMappingService.cs

```diff
@@ -5,22 +5,82 @@
 using System.Composition;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor;
 using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.Razor;
 using Microsoft.CodeAnalysis.Razor.DocumentMapping;
 using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
 using Microsoft.CodeAnalysis.Razor.Workspaces;
+using Microsoft.CodeAnalysis.Remote.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis.Remote.Razor.DocumentMapping;
 
 [Export(typeof(IDocumentMappingService)), Shared]
 [method: ImportingConstructor]
 internal sealed class RemoteDocumentMappingService(
     IFilePathService filePathService,
+    IProjectCollectionResolver projectCollectionResolver,
     ILoggerFactory loggerFactory)
     : AbstractDocumentMappingService(filePathService, loggerFactory.GetOrCreateLogger<RemoteDocumentMappingService>())
 {
-    protected override ValueTask<RazorCodeDocument?> TryGetCodeDocumentAsync(Uri razorDocumentUri, CancellationToken cancellationToken)
+    private readonly IProjectCollectionResolver _projectCollectionResolver = projectCollectionResolver;
+
+    public async Task<(Uri MappedDocumentUri, LinePositionSpan MappedRange)> MapToHostDocumentUriAndRangeAsync(
+        RemoteDocumentSnapshot originSnapshot,
+        Uri generatedDocumentUri,
+        LinePositionSpan generatedDocumentRange,
+        CancellationToken cancellationToken)
     {
-        throw new NotImplementedException();
+        var razorDocumentUri = FilePathService.GetRazorDocumentUri(generatedDocumentUri);
+
+        // For Html we just map the Uri, the range will be the same
+        if (FilePathService.IsVirtualHtmlFile(generatedDocumentUri))
+        {
+            return (razorDocumentUri, generatedDocumentRange);
+        }
+
+        // We only map from C# files
+        if (!FilePathService.IsVirtualCSharpFile(generatedDocumentUri))
+        {
+            return (generatedDocumentUri, generatedDocumentRange);
+        }
+
+        var razorFilePath = razorDocumentUri.GetAbsoluteOrUNCPath();
+        IDocumentSnapshot? razorDocumentSnapshot = null;
+
+        foreach (var project in _projectCollectionResolver.EnumerateProjects(originSnapshot))
+        {
+            if (project.TryGetDocument(razorFilePath, out razorDocumentSnapshot))
+            {
+                break;
+            }
+        }
+
+        if (razorDocumentSnapshot is not RemoteDocumentSnapshot targetDocumentSnapshot)
+        {
+            return (generatedDocumentUri, generatedDocumentRange);
```

Comment:
This feels Debug.Fail worthy. We're in remote, getting a snapshot, so it should always be a remote snapshot or things have gone really badly.

Reply:
FWIW, rather than `Debug.Fail`, I usually prefer `Assumed.Unreachable` to fail fast in cases like these. If there's no way this could happen debug, we shouldn't allow it to happen in release either.

Reply:
With the other changes you suggested, this check went away.